### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.loggers.neptune`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to `pytorch_lightning.loggers.neptune` for issue #13445

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by JustinGoheen** [comment]
blocked by https://github.com/Lightning-AI/lightning/pull/13689 for call to `_add_prefix` addressed in this [diff](https://github.com/Lightning-AI/lightning/pull/13689/files/28d5ec17c0fad18e878a2e3a3b548f152168100d#diff-a71f79d3349c77d9f589c4a098b211eae8102e787eee74bc5e8166953a4a468e) of linked PR

**Comment by otaj** [comment]
> blocked by #13689 for call to `_add_prefix` addressed in this [diff](https://github.com/Lightning-AI/lightning/pull/13689/files/28d5ec17c0fad18e878a2e3a3b548f152168100d#diff-a71f79d3349c77d9f589c4a098b211eae8102e787eee74bc5e8166953a4a468e) of linked PR

Hi @JustinGoheen, #13689 is already merged.

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/neptune.py:402]:
```suggestion
    def __setstate__(self, state: Dict[str, Any]) -> None:
```

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/neptune.py:396]:
```suggestion
    def __getstate__(self) -> Dict[str, Any]:
```

File: pyproject.toml
Comment (awaelchli) [inline pyproject.toml:64]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/neptune.py:338]: each call to this in the call hierarchy used a string. sanity checked the unpacking works with a str.

```python
LOGGER_JOIN_CHAR = "/"

def _construct_path_with_prefix(prefix, *keys: str) -> str:
    if prefix:
        return LOGGER_JOIN_CHAR.join([prefix, *keys])
    return LOGGER_JOIN_CHAR.join(keys)

_construct_path_with_prefix(None, "zed/quu")
Out: 'zed/quu'

_construct_path_with_prefix("root", "zed/quu")
Out: 'root/zed/quu'
```

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/neptune.py:622]: `_run_name` is annotated with Optional[None] in \_\_init_\_\, and is then set to a string in `._retrieve_run_data`

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/neptune.py:630]: `_run_sort_id` is annotated with Optional[None] in \_\_init_\_\, and is then set to a string in `._retrieve_run_data`.  mypy is incorrectly evaluating `_run_sort_id` as Optional[str]

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/neptune.py:657]: types `float` and `str` cause the following error despite `if torch.is_tensor(metric_value):` checking the type before executing `metric_value.cpu().detach()`

```sh
 Item ...  of "Union[Tensor, float, str]" has no attribute "cpu"
```

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/neptune.py:606]:
```suggestion
        structure_keys: List[str] = namespace.split(cls.LOGGER_JOIN_CHAR)
```

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [reply src/pytorch_lightning/loggers/neptune.py:657]: There is a check `torch.is_tensor(metric_value)` above, which means we know that it can only be a tensor. MyPy doesn't know this, but we can tell it this by 
`assert isinstance(metric_value, Tensor)`

IMO this should be done instead of the type ignore

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py:657]: I think future contributors/maintainers might see the assert and wonder why it was added after the is_tensor check, then remove it; whereas ignoring the error handles for mypy checks only, and does not introduce new code just to appease a mypy error.

In general, it would probably be good to set a convention on when, why, and how to change code or add ignore[union-attr] given the amount of optionals & unions in the code base.

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py:606]: when I make this change locally, the following error occurs:

```sh
src/pytorch_lightning/loggers/neptune.py:607: error: List item 1 has incompatible type "List[str]"; expected "Dict[Any, Any]"  [list-item]
```

Given the above, I updated ln 607 to 

```python
uploaded_models_dict: Dict[str, List[str]] = reduce(lambda d, k: d[k], [exp_structure, *structure_keys])
```

however, mypy then complained:

```sh
src/pytorch_lightning/loggers/neptune.py:607: error: Incompatible return value type (got "List[str]", expected "Dict[str, List[str]]")  [return-value]
src/pytorch_lightning/loggers/neptune.py:607: error: Invalid index type "Dict[str, List[str]]" for "Dict[str, List[str]]"; expected type "str"  [index]
src/pytorch_lightning/loggers/neptune.py:607: error: List item 1 has incompatible type "List[str]"; expected "Dict[str, List[str]]"  [list-item]
```

Likely because the lambda in reduce is using recursion to create the Dict `uploaded_models_dict`.

No errors occur when I leave the code as:

```python
def _get_full_model_names_from_exp_structure(cls, exp_structure: dict, namespace: str) -> Set[str]:
    """Returns all paths to properties which were already logged in `namespace`"""
    structure_keys: List = namespace.split(cls.LOGGER_JOIN_CHAR)
    uploaded_models_dict: Dict = reduce(lambda d, k: d[k], [exp_structure, *structure_keys])
    return set(cls._dict_paths(uploaded_models_dict))
```

Comment (awaelchli) [reply src/pytorch_lightning/loggers/neptune.py]: The reduction there is not compatible with the typing, so I changed it to a for loop. I'm not sure if it's "better" but this is the only way I found to make the types at least somewhat specific. Note, to properly type the dict there we would need some sort of recursive type definition, and I think that's not possible at the moment.

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [reply src/pytorch_lightning/loggers/neptune.py:658]: Actually you are right, we should change it to 

`if isinstance(metric_value, Tensor):`

this is even better. Then the assert is not needed and the ignore can be dropped.

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py]: updated in https://github.com/Lightning-AI/lightning/pull/13692/commits/4f049a97ba3d32ccb18b51320b85efecd67ad452

File: src/pytorch_lightning/loggers/neptune.py
Comment (otaj) [reply src/pytorch_lightning/loggers/neptune.py:622]: This should already be `str` by the time it's used, shouldn't it? If so, I'm voting for `assert` instead of `# type: ignore`

File: src/pytorch_lightning/loggers/neptune.py
Comment (otaj) [reply src/pytorch_lightning/loggers/neptune.py:630]: This should already be `str` by the time it's used, shouldn't it? If so, I'm voting for `assert` instead of `# type: ignore`

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py]: updated in https://github.com/Lightning-AI/lightning/pull/13692/commits/575f150fdb8cd7406d3a04f43a35be1fb1e63618

File: src/pytorch_lightning/loggers/neptune.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py]: updated in https://github.com/Lightning-AI/lightning/pull/13692/commits/575f150fdb8cd7406d3a04f43a35be1fb1e63618

Comment (otaj) [reply src/pytorch_lightning/loggers/neptune.py]: So, it seems this broke couple tests. Something fishy going on, I will debug it next week.

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py]: would it be more appropriate to use

```python
    @property
    def name(self) -> Optional[str]:
        """Return the experiment name or 'offline-name' when exp is run in offline mode."""
        if self._run_name is not None:
            return self._run_name

    @property
    def version(self) -> Optional[str]:
        """Return the experiment version.

        It's Neptune Run's short_id
        """
        if self._run_short_id is not None:
            return self._run_short_id
```

this passes mypy checks; I can at least push the change so that the other CI/CD checks can be compared across the two commits

Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/neptune.py]: implemented the above in https://github.com/Lightning-AI/lightning/pull/13692/commits/987337f647be9014f74c2efde6f8efde879b0de2

File: src/pytorch_lightning/loggers/neptune.py
Comment (otaj) [reply src/pytorch_lightning/loggers/neptune.py]: I took the liberty to update the code a bit to invoke `self.run` in order to fetch `self._run_name` and `self._run_short_id` to make the tests happy again

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/neptune.py:624]: @otaj 
The design here was intentional. As you can see in the base class, the name and version properties are Optional. This is because most loggers are not pickle friendly and we want to discourage accessing them in the main process before processes get spawned. This is why we create the experiments lazily, only when it is first needed. Hence, run name and run id would be None here because the experiment was not created yet.

Does it make sense?

Comment (otaj) [reply src/pytorch_lightning/loggers/neptune.py]: The idea makes sense. If the rest of the codebase is ok with`logger.name` being `None`, then so am I :joy: Thanks for the explanation.
I'm positive I saw a similar thing done in some other logger though, I'll have to look it up.

In the meantime, I'm reverting my two commits

**Review by awaelchli — APPROVED** [review]
Looks great!

File: src/pytorch_lightning/loggers/neptune.py
Comment (awaelchli) [reply src/pytorch_lightning/loggers/neptune.py]: Thanks. I know it's suboptimal but it is the only design that generalized well across all loggers for multiprocessing. So far. I definitely agree it would be nice if we could have no optionals there.

File: src/pytorch_lightning/loggers/neptune.py
Comment (rohitgr7) [inline src/pytorch_lightning/loggers/neptune.py:304]:
```suggestion
            assert self._run_instance is not None
            self._run_instance.wait()
            self._run_short_id = self._run_instance["sys/id"].fetch()
            self._run_name = self._run_instance["sys/name"].fetch()
```
I'd not change the logic.

**Review by rohitgr7 — APPROVED** [review]

**Pull request merged**
